<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Zilin Li" />

<meta name="date" content="2020-08-14" />

<title>Dynamic Scan Procedure for Detecting Rare-Variant Association Regions in Whole Genome Sequencing Studies</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>

</head>

<body>




<h1 class="title toc-ignore">Dynamic Scan Procedure for Detecting Rare-Variant Association Regions in Whole Genome Sequencing Studies</h1>
<h4 class="author"><em>Zilin Li</em></h4>
<h4 class="date"><em>2020-08-14</em></h4>



<p>The <code>SCANG</code> package accompanies the paper “Dynamic Scan Procedure for Detecting Rare-Variant Association Regions in Whole Genome Sequencing Studies” and is designed for detecting rare-variant association regions in whole genome sequencing studies.</p>
<p>Note: The results are depend on R 3.5.1.</p>
<div id="data-set-up-generating-signal-regions-covariates-family-structure" class="section level2">
<h2>Data Set-up: Generating signal regions, covariates, family-structure</h2>
<p>The genotype data was generated by simulating 10,000 samples for a 5 Mb region using COSI. We analyzed the 150,718 low frequency and rare variants (minor allele frequeny, MAF &lt; 0.05) in the sequence.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">library</span>(<span class="st">&quot;Matrix&quot;</span>)</a>
<a class="sourceLine" id="cb1-2" title="2"><span class="kw">library</span>(SCANG)</a>
<a class="sourceLine" id="cb1-3" title="3">genotype &lt;-<span class="st"> </span>example<span class="op">$</span>genotype</a>
<a class="sourceLine" id="cb1-4" title="4">maf &lt;-<span class="st"> </span>example<span class="op">$</span>maf</a>
<a class="sourceLine" id="cb1-5" title="5">snploc &lt;-<span class="st"> </span>example<span class="op">$</span>snploc</a>
<a class="sourceLine" id="cb1-6" title="6"><span class="kw">sum</span>((maf<span class="op">&gt;</span><span class="dv">0</span>)<span class="op">*</span>(maf<span class="op">&lt;</span><span class="fl">0.05</span>))</a></code></pre></div>
<pre><code>## [1] 150718</code></pre>
<p>We simulated 2,500 families and each family has 4 individuals. The kinship matrix is generated as follows:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="co">## kinship</span></a>
<a class="sourceLine" id="cb3-2" title="2"><span class="kw">library</span>(kinship2)</a></code></pre></div>
<pre><code>## Loading required package: quadprog</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="kw">library</span>(MASS)</a></code></pre></div>
<pre><code>## 
## Attaching package: &#39;MASS&#39;</code></pre>
<pre><code>## The following object is masked _by_ &#39;.GlobalEnv&#39;:
## 
##     genotype</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1">grid &lt;-<span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb8-2" title="2">Npercell &lt;-<span class="st"> </span><span class="dv">10000</span></a>
<a class="sourceLine" id="cb8-3" title="3">ndiv &lt;-<span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb8-4" title="4">vfam &lt;-<span class="st"> </span><span class="fl">0.5</span></a>
<a class="sourceLine" id="cb8-5" title="5">N &lt;-<span class="st"> </span><span class="kw">round</span>(grid<span class="op">*</span>grid<span class="op">*</span>Npercell<span class="op">/</span>ndiv)</a>
<a class="sourceLine" id="cb8-6" title="6"></a>
<a class="sourceLine" id="cb8-7" title="7">unitmat &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="fl">0.5</span>, <span class="dv">4</span>, <span class="dv">4</span>)</a>
<a class="sourceLine" id="cb8-8" title="8"><span class="kw">diag</span>(unitmat) &lt;-<span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb8-9" title="9">unitmat[<span class="dv">1</span>,<span class="dv">2</span>] &lt;-<span class="st"> </span>unitmat[<span class="dv">2</span>,<span class="dv">1</span>] &lt;-<span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb8-10" title="10">ped &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">famid =</span> <span class="kw">rep</span>(<span class="kw">as.integer</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">2500</span>), <span class="dt">each=</span><span class="dv">4</span>), <span class="dt">id =</span> <span class="kw">as.integer</span>(<span class="dv">1</span><span class="op">:</span>10000L), <span class="dt">fa =</span> <span class="kw">rep</span>(<span class="dv">0</span>, <span class="dv">10000</span>), <span class="dt">mo =</span> <span class="kw">rep</span>(<span class="dv">0</span>, <span class="dv">10000</span>))</a>
<a class="sourceLine" id="cb8-11" title="11"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2500</span>) {</a>
<a class="sourceLine" id="cb8-12" title="12">        ped<span class="op">$</span>fa[<span class="dv">4</span><span class="op">*</span>i<span class="op">-</span>(<span class="dv">0</span><span class="op">:</span><span class="dv">1</span>)] &lt;-<span class="st"> </span>ped<span class="op">$</span>id[<span class="dv">4</span><span class="op">*</span>i<span class="dv">-3</span>]</a>
<a class="sourceLine" id="cb8-13" title="13">    ped<span class="op">$</span>mo[<span class="dv">4</span><span class="op">*</span>i<span class="op">-</span>(<span class="dv">0</span><span class="op">:</span><span class="dv">1</span>)] &lt;-<span class="st"> </span>ped<span class="op">$</span>id[<span class="dv">4</span><span class="op">*</span>i<span class="dv">-2</span>]</a>
<a class="sourceLine" id="cb8-14" title="14">}</a>
<a class="sourceLine" id="cb8-15" title="15">kins &lt;-<span class="st"> </span><span class="kw">makekinship</span>(ped<span class="op">$</span>famid, ped<span class="op">$</span>id, ped<span class="op">$</span>fa, ped<span class="op">$</span>mo)</a></code></pre></div>
<p>We generated two covariates, one is continous, the other one is binary.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="kw">set.seed</span>(<span class="dv">168</span>)</a>
<a class="sourceLine" id="cb9-2" title="2">samplesize &lt;-<span class="st"> </span><span class="kw">dim</span>(genotype)[<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb9-3" title="3">X1 &lt;-<span class="st"> </span><span class="kw">rnorm</span>(samplesize,<span class="dv">0</span>,<span class="dv">1</span>)</a>
<a class="sourceLine" id="cb9-4" title="4">X2 &lt;-<span class="st"> </span><span class="kw">rbinom</span>(samplesize,<span class="dv">1</span>,<span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb9-5" title="5">epi &lt;-<span class="st"> </span><span class="kw">rnorm</span>(samplesize,<span class="dv">0</span>,<span class="dv">1</span>)</a></code></pre></div>
<p>We randomly selected two signal regions (genotype-phenotype association regions) across the 5 Mb region. The length of the signal region were ramdonly selected with lengths 3 kb, 4 kb, 5 kb and 6 kb.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="co">## number of signal regions</span></a>
<a class="sourceLine" id="cb10-2" title="2">n0 &lt;-<span class="st"> </span><span class="dv">2</span></a>
<a class="sourceLine" id="cb10-3" title="3"></a>
<a class="sourceLine" id="cb10-4" title="4"><span class="co">## generate signal region location</span></a>
<a class="sourceLine" id="cb10-5" title="5">pp &lt;-<span class="st"> </span><span class="kw">floor</span>(<span class="kw">dim</span>(genotype)[<span class="dv">2</span>]<span class="op">/</span>n0)</a>
<a class="sourceLine" id="cb10-6" title="6"></a>
<a class="sourceLine" id="cb10-7" title="7"><span class="co">## location od signal region</span></a>
<a class="sourceLine" id="cb10-8" title="8"><span class="co"># In snese of variants order</span></a>
<a class="sourceLine" id="cb10-9" title="9">sigloc &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rep</span>(<span class="dv">0</span>,<span class="dv">3</span><span class="op">*</span>n0),<span class="dt">ncol=</span><span class="dv">3</span>)</a>
<a class="sourceLine" id="cb10-10" title="10"><span class="co"># In sense of location</span></a>
<a class="sourceLine" id="cb10-11" title="11">sigloc_bp &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rep</span>(<span class="dv">0</span>,<span class="dv">3</span><span class="op">*</span>n0),<span class="dt">ncol=</span><span class="dv">3</span>)</a>
<a class="sourceLine" id="cb10-12" title="12"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n0)</a>
<a class="sourceLine" id="cb10-13" title="13">{</a>
<a class="sourceLine" id="cb10-14" title="14">  begnum &lt;-<span class="st"> </span>(i<span class="dv">-1</span>)<span class="op">*</span>pp <span class="op">+</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb10-15" title="15">  endnum &lt;-<span class="st"> </span>i<span class="op">*</span>pp <span class="op">-</span><span class="st"> </span><span class="dv">1000</span></a>
<a class="sourceLine" id="cb10-16" title="16">  sigloc[i,<span class="dv">1</span>] &lt;-<span class="st"> </span><span class="kw">sample</span>(begnum<span class="op">:</span>endnum,<span class="dv">1</span>)</a>
<a class="sourceLine" id="cb10-17" title="17">  sigloc_bp[i,<span class="dv">1</span>] &lt;-<span class="st"> </span>snploc<span class="op">$</span>CHROM_POS[sigloc[i,<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb10-18" title="18">  length_r &lt;-<span class="st"> </span><span class="kw">runif</span>(<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">1</span>)</a>
<a class="sourceLine" id="cb10-19" title="19">  sigloc_bp[i,<span class="dv">3</span>] &lt;-<span class="st"> </span>(length_r<span class="op">&lt;=</span><span class="dv">1</span><span class="op">/</span><span class="dv">4</span>)<span class="op">*</span><span class="dv">3000</span> <span class="op">+</span><span class="st"> </span>(length_r<span class="op">&gt;</span><span class="dv">1</span><span class="op">/</span><span class="dv">4</span>)<span class="op">*</span>(length_r<span class="op">&lt;=</span><span class="dv">2</span><span class="op">/</span><span class="dv">4</span>)<span class="op">*</span><span class="dv">4000</span>  <span class="op">+</span><span class="st"> </span>(length_r<span class="op">&gt;</span><span class="dv">2</span><span class="op">/</span><span class="dv">4</span>)<span class="op">*</span>(length_r<span class="op">&lt;=</span><span class="dv">3</span><span class="op">/</span><span class="dv">4</span>)<span class="op">*</span><span class="dv">5000</span> <span class="op">+</span><span class="st"> </span>(length_r<span class="op">&gt;</span><span class="dv">3</span><span class="op">/</span><span class="dv">4</span>)<span class="op">*</span><span class="dv">6000</span></a>
<a class="sourceLine" id="cb10-20" title="20">  sigloc_bp[i,<span class="dv">2</span>] &lt;-<span class="st"> </span>sigloc_bp[i,<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>sigloc_bp[i,<span class="dv">3</span>] <span class="op">-</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb10-21" title="21">  sigloc[i,<span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">which.max</span>(snploc<span class="op">$</span>CHROM_POS<span class="op">&gt;</span>sigloc_bp[i,<span class="dv">2</span>]) <span class="op">-</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb10-22" title="22">  <span class="cf">if</span>(sigloc[i,<span class="dv">2</span>] <span class="op">==</span><span class="st"> </span><span class="dv">0</span>)</a>
<a class="sourceLine" id="cb10-23" title="23">  {</a>
<a class="sourceLine" id="cb10-24" title="24">    sigloc[i,<span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">dim</span>(genotype)[<span class="dv">2</span>]</a>
<a class="sourceLine" id="cb10-25" title="25">  }</a>
<a class="sourceLine" id="cb10-26" title="26"></a>
<a class="sourceLine" id="cb10-27" title="27">  sigloc[i,<span class="dv">3</span>] &lt;-<span class="st"> </span>sigloc[i,<span class="dv">2</span>] <span class="op">-</span><span class="st"> </span>sigloc[i,<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb10-28" title="28">}</a></code></pre></div>
<p>Then the location of the two signal regions in sense of variants order were as follows (last column is the number of variants in the signal region):</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">sigloc</a></code></pre></div>
<pre><code>##        [,1]   [,2] [,3]
## [1,]  11889  12033  145
## [2,] 110249 110422  174</code></pre>
<p>The location of the two signal regions in sense of physical location were as follows (last column is the region length of the signal region):</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1">sigloc_bp</a></code></pre></div>
<pre><code>##         [,1]    [,2] [,3]
## [1,]  390258  395257 5000
## [2,] 3654712 3659711 5000</code></pre>
<p>We randomly select 10% of variants to be causal.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1">percen &lt;-<span class="st">  </span><span class="fl">0.10</span></a>
<a class="sourceLine" id="cb15-2" title="2"></a>
<a class="sourceLine" id="cb15-3" title="3">sigploc &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rep</span>(<span class="dv">0</span>,<span class="dv">28</span><span class="op">*</span>n0),<span class="dt">ncol=</span><span class="dv">28</span>)</a>
<a class="sourceLine" id="cb15-4" title="4">sigloctemp &lt;-<span class="st"> </span><span class="kw">c</span>()</a>
<a class="sourceLine" id="cb15-5" title="5"></a>
<a class="sourceLine" id="cb15-6" title="6"><span class="cf">for</span>(ii <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n0)</a>
<a class="sourceLine" id="cb15-7" title="7">{</a>
<a class="sourceLine" id="cb15-8" title="8">  maftemp &lt;-<span class="st"> </span>maf[(sigloc[ii,<span class="dv">1</span>]<span class="op">+</span><span class="dv">1</span>)<span class="op">:</span>(sigloc[ii,<span class="dv">2</span>]<span class="op">-</span><span class="dv">1</span>)]</a>
<a class="sourceLine" id="cb15-9" title="9">  mafid &lt;-<span class="st"> </span>(sigloc[ii,<span class="dv">1</span>]<span class="op">+</span><span class="dv">1</span>)<span class="op">:</span>(sigloc[ii,<span class="dv">2</span>]<span class="op">-</span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb15-10" title="10">  mafid &lt;-<span class="st"> </span>mafid[mafid<span class="op">&gt;</span><span class="dv">0</span>]</a>
<a class="sourceLine" id="cb15-11" title="11">  p0 &lt;-<span class="st"> </span><span class="kw">floor</span>(sigloc[ii,<span class="dv">3</span>]<span class="op">*</span>percen)</a>
<a class="sourceLine" id="cb15-12" title="12">  sigploc[ii,<span class="dv">1</span><span class="op">:</span>p0] &lt;-<span class="st"> </span><span class="kw">c</span>(sigloc[ii,<span class="dv">1</span>],<span class="kw">sort</span>(<span class="kw">sample</span>(mafid,p0<span class="dv">-2</span>,<span class="dt">replace=</span><span class="ot">FALSE</span>)),sigloc[ii,<span class="dv">2</span>])</a>
<a class="sourceLine" id="cb15-13" title="13">}</a>
<a class="sourceLine" id="cb15-14" title="14"></a>
<a class="sourceLine" id="cb15-15" title="15">sigloc &lt;-<span class="st"> </span><span class="kw">cbind</span>(sigloc,sigploc)</a></code></pre></div>
</div>
<div id="analyzing-the-simulated-data-using-scang" class="section level2">
<h2>Analyzing the simulated data using SCANG</h2>
<div id="unrelated-samples" class="section level3">
<h3>Unrelated samples</h3>
<p>We set the effect sizes as a decreasing function of MAFs, beta=0.18*log10(MAF). The coefficient for the causal variants are all positive. The phenotype of unrelated samples is generated through linear model.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1">protect &lt;-<span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb16-2" title="2">c0 &lt;-<span class="st"> </span><span class="fl">0.18</span></a>
<a class="sourceLine" id="cb16-3" title="3"></a>
<a class="sourceLine" id="cb16-4" title="4"><span class="co">## generate phenotype</span></a>
<a class="sourceLine" id="cb16-5" title="5">phenotype &lt;-<span class="st"> </span><span class="fl">0.5</span><span class="op">*</span>X1 <span class="op">+</span><span class="st"> </span><span class="fl">0.5</span><span class="op">*</span>X2</a>
<a class="sourceLine" id="cb16-6" title="6"><span class="cf">for</span>(ii <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n0)</a>
<a class="sourceLine" id="cb16-7" title="7">{</a>
<a class="sourceLine" id="cb16-8" title="8">  sigloctemp &lt;-<span class="st"> </span>sigloc[ii,<span class="dv">4</span><span class="op">:</span>(<span class="kw">dim</span>(sigloc)[<span class="dv">2</span>]<span class="op">-</span><span class="dv">1</span>)]</a>
<a class="sourceLine" id="cb16-9" title="9">  sigloctemp &lt;-<span class="st"> </span>sigloctemp[sigloctemp<span class="op">&gt;</span><span class="dv">0</span>]</a>
<a class="sourceLine" id="cb16-10" title="10">  <span class="co"># beta</span></a>
<a class="sourceLine" id="cb16-11" title="11">  beta &lt;-<span class="st"> </span>c0<span class="op">*</span><span class="kw">log</span>(maf[sigloctemp])<span class="op">/</span><span class="kw">log</span>(<span class="dv">10</span>)</a>
<a class="sourceLine" id="cb16-12" title="12">  betadir &lt;-<span class="st"> </span><span class="kw">rbinom</span>(<span class="kw">length</span>(beta),<span class="dv">1</span>,protect)<span class="op">*</span><span class="dv">2-1</span></a>
<a class="sourceLine" id="cb16-13" title="13">  beta &lt;-<span class="st"> </span>beta<span class="op">*</span>betadir</a>
<a class="sourceLine" id="cb16-14" title="14">  phenotype &lt;-<span class="st"> </span>phenotype <span class="op">+</span><span class="st"> </span>genotype[,sigloctemp]<span class="op">%*%</span>beta</a>
<a class="sourceLine" id="cb16-15" title="15">  phenotype &lt;-<span class="st"> </span><span class="kw">as.vector</span>(phenotype)</a>
<a class="sourceLine" id="cb16-16" title="16">}</a>
<a class="sourceLine" id="cb16-17" title="17"></a>
<a class="sourceLine" id="cb16-18" title="18">phenotype &lt;-<span class="st"> </span>phenotype <span class="op">+</span><span class="st"> </span>epi</a>
<a class="sourceLine" id="cb16-19" title="19">X &lt;-<span class="st"> </span><span class="kw">cbind</span>(X1,X2)</a>
<a class="sourceLine" id="cb16-20" title="20">phenotypedata &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">phenotype=</span>phenotype,<span class="dt">X=</span>X)</a></code></pre></div>
<p>We set the parameter of SCANG procedure. We set the smallest number of variants of searching windows as 70 and the maximum number of searching windows as 250. In this example, we consider the continuous trait and set the regression type as “gaussian”. We set the filtering threshold of SKAT as 2e-5, which indicates that SKAT p-values are calculated for regions whose p-value possibly is smaller than 2e-5.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1">Lmax &lt;-<span class="st"> </span><span class="dv">250</span>         </a>
<a class="sourceLine" id="cb17-2" title="2">Lmin &lt;-<span class="st"> </span><span class="dv">70</span>         </a>
<a class="sourceLine" id="cb17-3" title="3">family &lt;-<span class="st"> &quot;gaussian&quot;</span></a>
<a class="sourceLine" id="cb17-4" title="4">filter &lt;-<span class="st"> </span><span class="fl">2e-5</span></a>
<a class="sourceLine" id="cb17-5" title="5">f &lt;-<span class="st"> </span><span class="dv">0</span></a></code></pre></div>
<p>We apply SCANG to analyze the simulated data set. It takes around 1 hrs to analyze 5 Mb sequence on 10,000 unrelatedness individuals.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1"><span class="co"># obj_nullmodel &lt;- fit_null_glm_SCANG(phenotype~-1+X,data=phenotypedata,family=gaussian(link = &quot;identity&quot;))</span></a>
<a class="sourceLine" id="cb18-2" title="2"><span class="co"># res_lm &lt;- SCANG(genotype,obj_nullmodel,Lmin,Lmax,filter=filter,f=f)</span></a></code></pre></div>
<p>SCANG-O detected 2 significant signal regions, which are listed as follows:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1">res_lm<span class="op">$</span>SCANG_O_res</a></code></pre></div>
<pre><code>##          [,1]   [,2]   [,3]   [,4]
## [1,] 16.52558  11938  12077 0.0030
## [2,] 14.15696 110313 110382 0.0195</code></pre>
<p>Recall the true signal regions:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" title="1">sigloc[,<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>]</a></code></pre></div>
<pre><code>##        [,1]   [,2] [,3]   [,4]   [,5]   [,6]   [,7]   [,8]   [,9]  [,10]
## [1,]  11889  12033  145  11889  11893  11901  11913  11938  11953  11980
## [2,] 110249 110422  174 110249 110262 110288 110298 110307 110309 110312
##       [,11]  [,12]  [,13]  [,14]  [,15]  [,16]  [,17]  [,18]  [,19]  [,20]
## [1,]  11989  11999  12017  12019  12025  12029  12033      0      0      0
## [2,] 110314 110315 110336 110348 110360 110378 110388 110396 110404 110422</code></pre>
<p>The two detected region of SCANG-O is overlapped with the two true signal region, respectively. We next show the results of SCANG-S and SCANG-B.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" title="1">res_lm<span class="op">$</span>SCANG_S_res</a></code></pre></div>
<pre><code>##          [,1]   [,2]   [,3]   [,4]
## [1,] 17.21871  11938  12077 0.0005
## [2,] 13.09295 108584 108668 0.0160
## [3,] 11.90575 110294 110363 0.0450</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" title="1">res_lm<span class="op">$</span>SCANG_B_res</a></code></pre></div>
<pre><code>##         [,1]   [,2]   [,3]   [,4]
## [1,] 14.8472 110313 110382 0.0185</code></pre>
</div>
<div id="related-samples" class="section level3">
<h3>Related samples</h3>
<p>We set the effect sizes as a decreasing function of MAFs, beta=0.22*log10(MAF). The coefficient for the causal variants are all positive. The phenotype of unrelated samples is generated through linear model.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" title="1">protect &lt;-<span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb27-2" title="2">c0 &lt;-<span class="st"> </span><span class="fl">0.21</span></a>
<a class="sourceLine" id="cb27-3" title="3"><span class="co">## generate phenotype</span></a>
<a class="sourceLine" id="cb27-4" title="4">phenotype &lt;-<span class="st"> </span><span class="fl">0.5</span><span class="op">*</span>X1 <span class="op">+</span><span class="st"> </span><span class="fl">0.5</span><span class="op">*</span>X2</a>
<a class="sourceLine" id="cb27-5" title="5"><span class="cf">for</span>(ii <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n0)</a>
<a class="sourceLine" id="cb27-6" title="6">{</a>
<a class="sourceLine" id="cb27-7" title="7">  sigloctemp &lt;-<span class="st"> </span>sigloc[ii,<span class="dv">4</span><span class="op">:</span>(<span class="kw">dim</span>(sigloc)[<span class="dv">2</span>]<span class="op">-</span><span class="dv">1</span>)]</a>
<a class="sourceLine" id="cb27-8" title="8">  sigloctemp &lt;-<span class="st"> </span>sigloctemp[sigloctemp<span class="op">&gt;</span><span class="dv">0</span>]</a>
<a class="sourceLine" id="cb27-9" title="9">  <span class="co"># beta</span></a>
<a class="sourceLine" id="cb27-10" title="10">  beta &lt;-<span class="st"> </span>c0<span class="op">*</span><span class="kw">log</span>(maf[sigloctemp])<span class="op">/</span><span class="kw">log</span>(<span class="dv">10</span>)</a>
<a class="sourceLine" id="cb27-11" title="11">  betadir &lt;-<span class="st"> </span><span class="kw">rbinom</span>(<span class="kw">length</span>(beta),<span class="dv">1</span>,protect)<span class="op">*</span><span class="dv">2-1</span></a>
<a class="sourceLine" id="cb27-12" title="12">  beta &lt;-<span class="st"> </span>beta<span class="op">*</span>betadir</a>
<a class="sourceLine" id="cb27-13" title="13">  phenotype &lt;-<span class="st"> </span>phenotype <span class="op">+</span><span class="st"> </span>genotype[,sigloctemp]<span class="op">%*%</span>beta</a>
<a class="sourceLine" id="cb27-14" title="14">  phenotype &lt;-<span class="st"> </span><span class="kw">as.vector</span>(phenotype)</a>
<a class="sourceLine" id="cb27-15" title="15">}</a>
<a class="sourceLine" id="cb27-16" title="16"></a>
<a class="sourceLine" id="cb27-17" title="17"><span class="co">## random effect</span></a>
<a class="sourceLine" id="cb27-18" title="18">randomfam &lt;-<span class="st"> </span><span class="kw">mvrnorm</span>(N<span class="op">/</span><span class="dv">4</span>, <span class="kw">rep</span>(<span class="dv">0</span>, <span class="dv">4</span>), vfam <span class="op">*</span><span class="st"> </span>unitmat)</a>
<a class="sourceLine" id="cb27-19" title="19">randomfam &lt;-<span class="st"> </span><span class="kw">as.vector</span>(<span class="kw">t</span>(randomfam))</a>
<a class="sourceLine" id="cb27-20" title="20">id0 &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span>N</a>
<a class="sourceLine" id="cb27-21" title="21"></a>
<a class="sourceLine" id="cb27-22" title="22"></a>
<a class="sourceLine" id="cb27-23" title="23">phenotype &lt;-<span class="st"> </span>phenotype <span class="op">+</span><span class="st"> </span>randomfam <span class="op">+</span><span class="st"> </span>epi</a>
<a class="sourceLine" id="cb27-24" title="24">X &lt;-<span class="st"> </span><span class="kw">cbind</span>(X1,X2)</a>
<a class="sourceLine" id="cb27-25" title="25"></a>
<a class="sourceLine" id="cb27-26" title="26">phenotypedata_relatedness &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">phenotype=</span>phenotype,<span class="dt">X=</span>X,<span class="dt">id=</span>id0)</a></code></pre></div>
<div id="analyzing-data-using-sparse-grm" class="section level4">
<h4>Analyzing data using sparse GRM</h4>
<p>We apply SCANG to analyze the simulated data set. Note that the GRM is sparse matrix.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" title="1"><span class="co"># obj_nullmodel &lt;- fit_null_glmmkin_SCANG(phenotype ~ -1+X, data=phenotypedata_relatedness, kins=kins, id=&quot;id&quot;,family=gaussian(link=&quot;identity&quot;))</span></a>
<a class="sourceLine" id="cb28-2" title="2"></a>
<a class="sourceLine" id="cb28-3" title="3"><span class="co"># res_lmm &lt;- SCANG(genotype=genotype,obj_nullmodel=obj_nullmodel,Lmin=Lmin,Lmax=Lmax,filter=filter,f=f)</span></a></code></pre></div>
<p>SCANG-O detected 2 significant signal regions, which are listed as follows:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" title="1">res_lmm<span class="op">$</span>SCANG_O_res</a></code></pre></div>
<pre><code>##          [,1]   [,2]   [,3]   [,4]
## [1,] 14.38947 110312 110401 0.0180
## [2,] 13.37193  11938  12077 0.0385</code></pre>
<p>The two detected region is overlapped with the two true signal region, respectively. We next show the results of SCANG-S and SCANG-B.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" title="1">res_lmm<span class="op">$</span>SCANG_S_res</a></code></pre></div>
<pre><code>##         [,1]  [,2]  [,3]  [,4]
## [1,] 14.0651 11938 12077 0.008</code></pre>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" title="1">res_lmm<span class="op">$</span>SCANG_B_res</a></code></pre></div>
<pre><code>##          [,1]   [,2]   [,3]   [,4]
## [1,] 15.07736 110312 110401 0.0145</code></pre>
</div>
<div id="analyzing-data-using-dense-grm" class="section level4">
<h4>Analyzing data using dense GRM</h4>
<p>We first transform the GRM into dense matrix.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" title="1">kins &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(kins)</a></code></pre></div>
<p>We then apply SCANG to analyze the simulated data set.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" title="1"><span class="co"># obj_nullmodel &lt;- fit_null_glmmkin_SCANG(phenotype ~ -1+X, data=phenotypedata_relatedness, kins=kins, id=&quot;id&quot;,family=gaussian(link=&quot;identity&quot;),use_sparse = FALSE)</span></a>
<a class="sourceLine" id="cb36-2" title="2"><span class="co"># res_lmm_dense &lt;- SCANG(genotype,obj_nullmodel,Lmin,Lmax,filter=filter,f=f)</span></a></code></pre></div>
<p>SCANG-O detected 2 significant signal regions, which are listed as follows:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" title="1">res_lmm_dense<span class="op">$</span>SCANG_O_res</a></code></pre></div>
<pre><code>##          [,1]   [,2]   [,3]   [,4]
## [1,] 14.38947 110312 110401 0.0175
## [2,] 13.37193  11938  12077 0.0450</code></pre>
<p>The two detected region is overlapped with the two true signal region, respectively. We next show the results of SCANG-S and SCANG-B.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" title="1">res_lmm_dense<span class="op">$</span>SCANG_S_res</a></code></pre></div>
<pre><code>##          [,1]  [,2]  [,3]   [,4]
## [1,] 14.06501 11938 12077 0.0085</code></pre>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" title="1">res_lmm_dense<span class="op">$</span>SCANG_B_res</a></code></pre></div>
<pre><code>##          [,1]   [,2]   [,3]   [,4]
## [1,] 15.07736 110312 110401 0.0155</code></pre>
<p>The results of using dense GRM and sparse GRM are nearly the same.</p>
</div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
